content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx

```diff
@@ -0,0 +1,401 @@
+
+You've also defined a `WithFullName<User>` return type, which takes whatever `User` is and tacks on a `fullName` string attribute.
+
+With your function, any object that contains `firstName` and `lastName` keys can compute a `fullName`. Pretty neat, right?
+
+## Excluding **Sensitive** Fields
+
+Very often, you don't want to include the **whole response** back from the database. In the signup function above, we've included the `password` in the response. Let's exclude this sensitive piece of data from the response.
+
+### Exclude Password in Javascript
+
+Once again, we'll start with Javascript so you can first focus on the logic. You first want to create an `exclude` helper function that takes a user and returns the user without certain fields:
+
+```tsx file=index.ts
+// Remove keys from user
+function exclude(user, ...keys) {
+  for (let key of keys) {
+    delete user[key]
+  }
+  return user
+}
+
+function main() {
+  const user = await prisma.user.findUnique({ where: 1 })
+  const userWithoutPassword = exclude(user, 'password')
+}
+```
+
+### Exclude Password in Typescript
+
+Now let's make this `exclude` function type-safe. You first want to make sure that the keys you provide exist on the user and then you want to return a user that doesn't contain those keys.
+
+Here's what that would look like in Typescript:
+
+```tsx file=index.ts highlight=1-4;normal
+function exclude<User, Key extends keyof User>(
+  user: User,
+  ...keys: Key[]
+): Omit<User, Key> {
+  for (let key of keys) {
+    delete user[key]
+  }
+  return user
+}
+
+function main() {
+  const user = await prisma.user.findUnique({ where: 1 })
+  const userWithoutPassword = exclude(user, 'password')
+}
+```
+
+In this example, we've provided two generics `User` and `Key`. The `Key` generic is defined as the keys of a `User` (e.g. email, password, firstName, etc.). These generics flow through the logic, returning a User that omits the list of keys provided.
+
+## Organizing the Prisma Client **into Models**
+
+In the examples above, you've created helper functions that weave the type-safety deeper through your application. Creating helper functions works well when your application is small, but as your application grows, you may find an inconsistent use of these helper functions strewn across your codebase.
+
+One technique for consistently applying and organizing these helper functions is to group them in a class.
+
+### Wrapping the Model in a Class
+
+The most obvious way to wrap the Prisma Client is by wrapping it in a `Model` class:
+
+```tsx highlight=4,23,28-30;normal
+import { PrismaClient } from '@prisma/client'
+
+class Model {
+  constructor(private readonly model: PrismaClient['user']) {}
+
+  /**
+   * Signup the first user and create a new team of one. Return the User with
+   * a full name and without a password
+   */
+  async signup(input: Signup) {
+    assert(input, Signup)
+    const team = await prisma.team.create({
```

Comment:
I guess this should use `model` instead of `prisma` directly, or otherwise there is no point in passing the prisma client as a constructor argument, or is there?